docs/software/ml/pytorch.md

```diff
@@ -0,0 +1,394 @@
+    #SBATCH --view=default
+
+    #################################
+    # OpenMP environment variables #
+    #################################
+    export OMP_NUM_THREADS=$SLURM_CPUS_PER_TASK  # (2)!
+
+    #################################
+    # PyTorch environment variables #
+    #################################
+    export MASTER_ADDR=$(hostname) # (3)!
+    export MASTER_PORT=6000
+    export WORLD_SIZE=$SLURM_NPROCS
+    export TORCH_NCCL_ASYNC_ERROR_HANDLING=1 # (4)!
+    export TRITON_HOME=/dev/shm/ # (5)!
+
+    #################################
+    # MPICH environment variables   #
+    #################################
+    export MPICH_GPU_SUPPORT_ENABLED=0 # (6)!
+
+    #################################
+    # CUDA environment variables    #
+    #################################
+    export CUDA_CACHE_DISABLE=1 # (7)!
+
+    ############################################
+    # NCCL and Fabric environment variables    #
+    ############################################
+    export NCCL_NET="AWS Libfabric" # (8)!
+    export NCCL_NET_GDR_LEVEL=PHB
+    export NCCL_CROSS_NIC=1
+    export FI_CXI_DISABLE_HOST_REGISTER=1
+    export FI_MR_CACHE_MONITOR=userfaultfd
+    export FI_CXI_DEFAULT_CQ_SIZE=131072
+    export FI_CXI_DEFAULT_TX_SIZE=32768
+    export FI_CXI_RX_MATCH_MODE=software
+
+    # (9)!
+    # (10)!
+    srun bash -c "
+        export RANK=\$SLURM_PROCID
+        export LOCAL_RANK=\$SLURM_LOCALID
+        . ./my-venv/bin/activate
+        python myscript.py
+    "
+    ```
+
+    1. The `--uenv` option is used to specify the uenv to use for the job.
+       The `--view=default` option is used to load all the packages provided by the uenv.
+    2. Only set `OMP_NUM_THREADS` if you are using OpenMP in your code.
+    3. These variables are used by PyTorch to initialize the distributed backend.
+       The `MASTER_ADDR` and `MASTER_PORT` variables are used to determine the address and port of the master node.
+       Additionally we also need `RANK` and `LOCAL_RANK` but these must be set per-process, see below.
+    4. Enable more graceful exception handling, see [PyTorch documentation](https://pytorch.org/docs/stable/torch_nccl_environment_variables.html)
+    5. Set the Trition home to a local path (e.g. `/dev/shm`) to avoid writing to the (distributed) file system.
+       This is important for performance, as writing to the Lustre file system can be slow due to the amount of small files and potentially many processes accessing it.
+    6. Disable GPU support in MPICH, as it [can lead to deadlocks](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/mpi.html#inter-gpu-communication-with-cuda-aware-mpi) when using together with nccl.
+    6. Avoid writing JITed binaries to the (distributed) file system, which could lead to performance issues.
+    7. These variables should always be set for correctness and optimal performance when using NCCL, see [the detailed explanation][ref-communication-nccl].
```

Comment:
as I said this is great and can be merged as it is, but I just wonder if there was a way to set these vars in uenv like we do with the CE hook?

Reply:
yes, there is. But it requires maintaining a custom spack package for aws-ofi-nccl in alps-cluster-config. I will do this as soon as I have confidence in all the version combinations and settings. Then all uenv's would need to be recompiled and redeployed.